modules/vmfcore/src/statistics.cpp

```diff
@@ -0,0 +1,1276 @@
+                }
+            }
+            else
+            {
+                m_metadataDesc = nullptr;
+                m_fieldDesc = FieldDesc();
+            }
+        }
+    MetadataStream* getStream() const
+        { return m_pMetadataStream; }
+
+private:
+    std::string m_name;
+    std::string m_schemaName;
+    std::string m_metadataName;
+    std::shared_ptr< MetadataDesc > m_metadataDesc;
+    std::string m_fieldName;
+    FieldDesc m_fieldDesc;
+    std::string m_opName;
+    MetadataStream* m_pMetadataStream;
+};
+
+StatField::StatField(
+        const std::string& name,
+        const std::string& schemaName,
+        const std::string& metadataName,
+        const std::string& fieldName,
+        const std::string& opName )
+    : m_desc( new StatFieldDesc( name, schemaName, metadataName, fieldName, opName ))
+    , m_op( StatOpFactory::create( opName ))
+    , m_state( StatState::UpToDate )
+    , m_isActive( false )
+{
+}
+
+StatField::StatField( const StatField& other )
+    : m_desc( new StatFieldDesc( *other.m_desc ))
+    , m_op( (other.m_op != nullptr) ? StatOpFactory::create( other.m_op->name() ) : nullptr )
+    , m_state( other.m_state )
+    , m_isActive( other.m_isActive )
+{
+}
+
+StatField::StatField( StatField&& other )
+    : m_desc( std::move( other.m_desc ))
+    , m_op( nullptr )
+    , m_state( std::move( other.m_state ))
+    , m_isActive( std::move( other.m_isActive ))
+{
+    std::swap( m_op, other.m_op );
+}
+
+StatField::StatField()
+    : m_desc()
+    , m_op( nullptr )
+    , m_state( StatState::UpToDate )
+    , m_isActive( false )
+{
+}
+
+StatField::~StatField()
+{
+    delete m_op; m_op = nullptr;
+}
+
+StatField& StatField::operator=( const StatField& other )
+{
+    setStream( nullptr );
+
+    m_desc.reset( new StatFieldDesc( *other.m_desc ));
+
+    delete m_op;
+    m_op = (other.m_op != nullptr) ? StatOpFactory::create( other.m_op->name() ) : nullptr;
+
+    m_state = other.m_state;
+    m_isActive = other.m_isActive;
+
+    setStream( other.getStream() );
+
+    return *this;
+}
+
+StatField& StatField::operator=( StatField&& other )
+{
+    setStream( nullptr );
+
+    m_desc = std::move( other.m_desc );
+
+    delete m_op;
+    m_op = nullptr;
+    std::swap( m_op, other.m_op );
+
+    m_state = other.m_state;
+    m_isActive = other.m_isActive;
+
+    setStream( other.getStream() );
+
+    return *this;
+}
+
+StatState::Type StatField::handle( std::shared_ptr< Metadata > metadata )
+{
+    const std::shared_ptr< MetadataDesc > metadataDesc = metadata->getDesc();
+    if( metadataDesc && (this->getMetadataDesc() == metadataDesc ))
+    {
+        const std::string& fieldName = this->getFieldName();
+        Metadata::iterator it = metadata->findField( fieldName );
+        if( it != metadata->end() )
+        {
+            updateState( m_op->handle( StatAction::Add, *it ));
+        }
+    }
+    return getState();
+}
+
+void StatField::update( bool doRescan )
```

Comment:
Changed implementation of the update/rescan subsystem with a focus on Stat side, and outer loop over metadata set.